In [1]:
import pandas as pd
from pathlib import Path

In [2]:
RAW_DATA_PATH = Path("../data/raw")

files = list(RAW_DATA_PATH.glob("*.csv"))

for file in files:
    print(file.name)

olist_sellers_dataset.csv
product_category_name_translation.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv


In [5]:
for file in files:
    df = pd.read_csv(file)

    print("=" * 80)
    print(f"File: {file.name}")
    print(f"Shape: {df.shape}")
    print("\nColumns:")
    print(df.columns.tolist())
    print()

File: olist_sellers_dataset.csv
Shape: (3095, 4)

Columns:
['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']

File: product_category_name_translation.csv
Shape: (71, 2)

Columns:
['product_category_name', 'product_category_name_english']

File: olist_orders_dataset.csv
Shape: (99441, 8)

Columns:
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

File: olist_order_items_dataset.csv
Shape: (112650, 7)

Columns:
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

File: olist_customers_dataset.csv
Shape: (99441, 5)

Columns:
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

File: olist_geolocation_dataset.csv
Shape: (1000163, 5)

Columns:
['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city

In [4]:
for file in files:
    df = pd.read_csv(file)

    print("=" * 80)
    print(f"File: {file.name}")
    print(f"Rows: {len(df):,}")
    print(f"Columns: {len(df.columns)}")
    print(f"Duplicate rows: {df.duplicated().sum():,}")

    missing = df.isnull().sum()
    missing = missing[missing > 0]

    print("\nMissing values:")
    if len(missing) == 0:
        print("None")
    else:
        print(missing)

    print()

File: olist_sellers_dataset.csv
Rows: 3,095
Columns: 4
Duplicate rows: 0

Missing values:
None

File: product_category_name_translation.csv
Rows: 71
Columns: 2
Duplicate rows: 0

Missing values:
None

File: olist_orders_dataset.csv
Rows: 99,441
Columns: 8
Duplicate rows: 0

Missing values:
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
dtype: int64

File: olist_order_items_dataset.csv
Rows: 112,650
Columns: 7
Duplicate rows: 0

Missing values:
None

File: olist_customers_dataset.csv
Rows: 99,441
Columns: 5
Duplicate rows: 0

Missing values:
None

File: olist_geolocation_dataset.csv
Rows: 1,000,163
Columns: 5
Duplicate rows: 261,831

Missing values:
None

File: olist_order_payments_dataset.csv
Rows: 103,886
Columns: 5
Duplicate rows: 0

Missing values:
None

File: olist_order_reviews_dataset.csv
Rows: 99,224
Columns: 7
Duplicate rows: 0

Missing values:
review_comment_title      87656
review_comment_message    58247
dtyp

In [6]:
profile = []

for file in files:
    df = pd.read_csv(file)

    profile.append({
        "table": file.stem,
        "rows": len(df),
        "columns": len(df.columns),
        "duplicate_rows": df.duplicated().sum(),
        "missing_cells": df.isna().sum().sum(),
        "missing_columns": (df.isna().sum() > 0).sum()
    })

profile_df = pd.DataFrame(profile)

profile_df.sort_values("rows", ascending=False)

,table,rows,columns,duplicate_rows,missing_cells,missing_columns
5,olist_geolocation_dataset,1000163,5,261831,0,0
3,olist_order_items_dataset,112650,7,0,0,0
6,olist_order_payments_dataset,103886,5,0,0,0
2,olist_orders_dataset,99441,8,0,4908,3
4,olist_customers_dataset,99441,5,0,0,0
7,olist_order_reviews_dataset,99224,7,0,145903,2
8,olist_products_dataset,32951,9,0,2448,8
0,olist_sellers_dataset,3095,4,0,0,0
1,product_category_name_translation,71,2,0,0,0


In [7]:
core_tables = [
    "olist_orders_dataset.csv",
    "olist_customers_dataset.csv",
    "olist_order_items_dataset.csv",
    "olist_products_dataset.csv",
    "olist_sellers_dataset.csv"
]

for table in core_tables:
    df = pd.read_csv(RAW_DATA_PATH / table)

    print("=" * 80)
    print(table)
    print(df.columns.tolist())
    display(df.head(3))

olist_orders_dataset.csv
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00


olist_customers_dataset.csv
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP


olist_order_items_dataset.csv
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87


olist_products_dataset.csv
['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0


olist_sellers_dataset.csv
['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ


In [8]:
orders = pd.read_csv(RAW_DATA_PATH / "olist_orders_dataset.csv")
customers = pd.read_csv(RAW_DATA_PATH / "olist_customers_dataset.csv")
order_items = pd.read_csv(RAW_DATA_PATH / "olist_order_items_dataset.csv")
products = pd.read_csv(RAW_DATA_PATH / "olist_products_dataset.csv")
sellers = pd.read_csv(RAW_DATA_PATH / "olist_sellers_dataset.csv")

print(
    "orders:",
    len(orders),
    orders["order_id"].nunique()
)

print(
    "customers:",
    len(customers),
    customers["customer_id"].nunique()
)

print(
    "products:",
    len(products),
    products["product_id"].nunique()
)

print(
    "sellers:",
    len(sellers),
    sellers["seller_id"].nunique()
)

print(
    "order_items:",
    len(order_items),
    order_items[["order_id", "order_item_id"]].drop_duplicates().shape[0]
)

orders: 99441 99441
customers: 99441 99441
products: 32951 32951
sellers: 3095 3095
order_items: 112650 112650


In [9]:
print("Unique customers (customer_id):",
      customers["customer_id"].nunique())

print("Unique real customers (customer_unique_id):",
      customers["customer_unique_id"].nunique())

print("Unique orders:",
      orders["order_id"].nunique())

print("Unique orders in order_items:",
      order_items["order_id"].nunique())

print("Unique products sold:",
      order_items["product_id"].nunique())

print("Unique sellers active:",
      order_items["seller_id"].nunique())

Unique customers (customer_id): 99441
Unique real customers (customer_unique_id): 96096
Unique orders: 99441
Unique orders in order_items: 98666
Unique products sold: 32951
Unique sellers active: 3095


In [10]:
print(
    "Orders without customer:",
    (~orders["customer_id"].isin(customers["customer_id"])).sum()
)

print(
    "Order items without order:",
    (~order_items["order_id"].isin(orders["order_id"])).sum()
)

print(
    "Order items without product:",
    (~order_items["product_id"].isin(products["product_id"])).sum()
)

print(
    "Order items without seller:",
    (~order_items["seller_id"].isin(sellers["seller_id"])).sum()
)

Orders without customer: 0
Order items without order: 0
Order items without product: 0
Order items without seller: 0


In [11]:
orders.isna().sum().sort_values(ascending=False)

order_delivered_customer_date    2965
order_delivered_carrier_date     1783
order_approved_at                 160
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_estimated_delivery_date       0
dtype: int64

In [12]:
orders.groupby("order_status")[
    [
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date"
    ]
].apply(lambda x: x.isna().sum())

,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date
order_status,,,
approved,0,2,2
canceled,141,550,619
created,5,5,5
delivered,14,2,8
invoiced,0,314,314
processing,0,301,301
shipped,0,0,1107
unavailable,0,609,609


订单数
COUNT(DISTINCT order_id)

订单客户记录数
COUNT(DISTINCT customer_id)

真实独立用户数
COUNT(DISTINCT customer_unique_id)

没有 item 的订单很可能集中在 canceled、unavailable 等状态

customer_unique_id
      │
      │ 1 : N   同一个真实用户可以多次下单
      ▼
customers
PK customer_id
      │
      │ 1 : 1   每个 customer_id 对应一笔订单
      ▼
orders
PK order_id
      │
      │ 1 : N
      ▼
order_items
(order_id + order_item_id)
    │                 │
product_id         seller_id
    ▼                 ▼
 products           sellers

In [13]:
orders_without_items = orders[
    ~orders["order_id"].isin(order_items["order_id"])
]

print("Orders without items:", len(orders_without_items))

display(
    orders_without_items["order_status"]
    .value_counts()
    .rename_axis("order_status")
    .reset_index(name="orders")
)

Orders without items: 775


,order_status,orders
0,unavailable,603
1,canceled,164
2,created,5
3,invoiced,2
4,shipped,1


orders 中存在订单，但 order_items 中没有商品记录，不一定是数据错误，而可能是订单根本没有完成正常商品履约流程。shipped = 1，订单已经 shipped 却没有 item，比较异常，记录在Data Quality Notes

In [14]:
product_missing = (
    products.isna()
    .sum()
    .sort_values(ascending=False)
)

product_missing[product_missing > 0]

product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [15]:
orders_without_items = orders[
    ~orders["order_id"].isin(order_items["order_id"])
]

orders_without_items["order_status"].value_counts()

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

## Data Profiling Summary

- 99,441 orders and 96,096 unique customers were identified.
- All core primary-key candidates are unique.
- No orphan foreign-key records were found across the five core tables.
- 775 orders have no associated order items, primarily unavailable or canceled orders.
- Missing order timestamps are largely explained by order lifecycle status and should not be removed blindly.
- 610 products lack category/content metadata, while only 2 products lack physical dimensions.
- Raw data will remain unchanged; cleaning outputs will be stored under `data/processed/`.

In [16]:
reviews = pd.read_csv(
    RAW_DATA_PATH / "olist_order_reviews_dataset.csv"
)

print("Rows:", len(reviews))
print("Unique review_id:", reviews["review_id"].nunique())
print("Unique order_id:", reviews["order_id"].nunique())
print("Missing review_score:", reviews["review_score"].isna().sum())

print("\nReviews per order:")
print(
    reviews.groupby("order_id")
    .size()
    .value_counts()
    .sort_index()
)

Rows: 99224
Unique review_id: 98410
Unique order_id: 98673
Missing review_score: 0

Reviews per order:
1    98126
2      543
3        4
Name: count, dtype: int64
